src/ray/raylet/scheduling/cluster_task_manager_test.cc

```diff
@@ -270,6 +282,70 @@ TEST_F(ClusterTaskManagerTest, BasicTest) {
   AssertNoLeaks();
 }
 
+TEST_F(ClusterTaskManagerTest, DispatchQueueNonBlockingTest) {
+  /*
+    Test that if no worker is available for the first task in a dispatch
+    queue (because the runtime env in the task spec doesn't match any
+    available worker), other tasks in the dispatch queue can still be scheduled.
+    https://github.com/ray-project/ray/issues/16226
+   */
+
+  // Use the same required_resources for all tasks so they end up in the same queue.
+  const std::unordered_map<std::string, double> required_resources = {
+      {ray::kCPU_ResourceLabel, 4}};
+
+  std::string serialized_runtime_env_A = "mock_env_A";
+  Task task_A = CreateTask(required_resources, /*num_args=*/0, /*args=*/{},
+                           serialized_runtime_env_A);
+  rpc::RequestWorkerLeaseReply reply_A;
+  bool callback_occurred = false;
+  bool *callback_occurred_ptr = &callback_occurred;
+  auto callback = [callback_occurred_ptr](Status, std::function<void()>,
+                                          std::function<void()>) {
+    *callback_occurred_ptr = true;
+  };
+
+  std::string serialized_runtime_env_B = "mock_env_B";
+  Task task_B_1 = CreateTask(required_resources, /*num_args=*/0, /*args=*/{},
+                             serialized_runtime_env_B);
+  Task task_B_2 = CreateTask(required_resources, /*num_args=*/0, /*args=*/{},
+                             serialized_runtime_env_B);
+  rpc::RequestWorkerLeaseReply reply_B_1;
+  rpc::RequestWorkerLeaseReply reply_B_2;
+  auto empty_callback = [](Status, std::function<void()>, std::function<void()>) {};
+
+  // Ensure task_A is not at the front of the queue.
+  task_manager_.QueueAndScheduleTask(task_B_1, &reply_B_1, empty_callback);
+  task_manager_.QueueAndScheduleTask(task_A, &reply_A, callback);
+  task_manager_.QueueAndScheduleTask(task_B_2, &reply_B_2, empty_callback);
+
+  // Push a worker that can only run task A.
+  const WorkerCacheKey env_A = {/*override_environment_variables=*/{},
+                                serialized_runtime_env_A};
+  const int runtime_env_hash_A = env_A.IntHash();
+  std::shared_ptr<MockWorker> worker_A =
+      std::make_shared<MockWorker>(WorkerID::FromRandom(), 1234, runtime_env_hash_A);
+  pool_.PushWorker(std::static_pointer_cast<WorkerInterface>(worker_A));
+  ASSERT_EQ(pool_.workers.size(), 1);
+
+  // Check we can schedule task A, even though task B is at the front of the queue
+  // and no workers are available for task B.
+  task_manager_.ScheduleAndDispatchTasks();
+
+  ASSERT_TRUE(callback_occurred);
+  ASSERT_EQ(leased_workers_.size(), 1);
+  ASSERT_EQ(pool_.workers.size(), 0);
+  ASSERT_EQ(node_info_calls_, 0);
+
+  Task finished_task;
+  task_manager_.TaskFinished(leased_workers_.begin()->second, &finished_task);
+  ASSERT_EQ(finished_task.GetTaskSpecification().TaskId(),
+            task_A.GetTaskSpecification().TaskId());
+
+  // task_B_1 and task_B_2 remain in the dispatch queue, so don't call AssertNoLeaks().
+  // AssertNoLeaks();
```

Comment:
Is it okay to not have AssertNoLeaks() at the end of this test?  I'm not sure how the tests in this file are isolated from one another, given that they share the same `task_manager_` and so on.  In any case, this test didn't seem to interfere with any of the other tests.

Reply:
I believe the full fixture should be re-run with every individual test case so we should reconstruct the task manager and it should be ok.

Reply:
We should uncomment this assert before merging right?

Reply:
Sorry, do you mean delete lines 345 and 346?  If we uncomment it, it will fail because there are tasks left in the dispatch queue.

Reply:
sorry, i'm dumb